fcitx5-fbterm.cpp

```diff
@@ -0,0 +1,353 @@
+            char* str = keysym_to_term_string(linux_keysym, down);
+            if (str)
+                put_im_text(str, strlen(str));
+        }
+
+        state = calculate_modifiers(state, keysym, down);
+    }
+}
+
+static void cursor_pos_changed(unsigned x, unsigned y) {
+    cursorx = x;
+    cursory = y;
+    im_show(-1);
+}
+
+static void update_fbterm_info(Info* info) {
+    fbtermInfo = *info;
+}
+
+static ImCallbacks cbs = {
+        im_active,  // .active
+        im_deactive,// .deactive
+        im_show,
+        im_hide,
+        process_raw_key,   // .send_key
+        cursor_pos_changed,// .cursor_position
+        update_fbterm_info,// .fbterm_info
+        update_term_mode   // .term_mode
+};
+
+void fcitx_fbterm_connect_cb(FcitxGClient*, void*) {
+    g_assert(fcitx_g_client_is_valid(client));
+    fcitx_g_client_set_capability(client, static_cast<guint64>(fcitx::CapabilityFlag::ClientSideInputPanel));
+}
+
+void fcitx_fbterm_commit_string_cb(FcitxGClient*, char* str, void*) {
+    put_im_text(str, strlen(str));
+}
+
+void fcitx_fbterm_current_im_cb(FcitxGClient*, char* name, char* uniqe_name, char* lang_code, gint type, void*) {
+    state = fcitx::KeyState::NoState;
+    //    snprintf(imname, BUF_SIZE, "%s", name);
+    //    im_show(-1);
+}
+
+void fcitx_fbterm_update_client_side_ui_cb_parsed(const char* preedit, int _cursorPos, const char* auxUp,
+                                                  const char* auxDown, const vector<FcitxGCandidateItem*>& candidates,
+                                                  int highlight, bool hasPrev, bool hasNext) {
+    cursorPos = _cursorPos;
+    snprintf(textup, BUF_SIZE, "%s%s", auxUp, preedit);
+    auto down = string();
+    down.append(auxDown);
+    for (int i = 0; i < candidates.size(); i++) {
+        down.append(i == highlight ? "*" : " ");
+        auto idx = to_string(i <= 8 ? i + 1 : i - 9);
+        down.append(idx + ".");
+        down.append(candidates[i]->candidate);
+    }
+    snprintf(textdown, BUF_SIZE, "%s", down.c_str());
+    textup[BUF_SIZE - 1] = textdown[BUF_SIZE - 1] = '\0';
+    auxUpWidth = text_width((char*)auxUp);
+    im_show(-1);
+}
+
+void fcitx_fbterm_update_client_side_ui_cb(FcitxGClient*, GPtrArray* preedit, int _cursorPos, GPtrArray* auxUp,
+                                           GPtrArray* auxDown, GPtrArray* candidates, int highlight,
+                                           int /* layoutHint */, gboolean hasPrev, gboolean hasNext,
+                                           void*) {
+    auto p1 = preedit->len > 0 ? ((FcitxGPreeditItem*)g_ptr_array_index(preedit, 0))->string : "";
+    auto p2 = _cursorPos;
+    auto p3 = auxUp->len > 0 ? ((FcitxGPreeditItem*)g_ptr_array_index(auxUp, 0))->string : "";
+    auto p4 = auxDown->len > 0 ? ((FcitxGPreeditItem*)g_ptr_array_index(auxDown, 0))->string : "";
+    auto p5 = vector<FcitxGCandidateItem*>();
+    auto p6 = highlight;
+    auto p7 = hasPrev;
```

Comment:
Can we have variable name more meaningful?

Also, hasNext hasPrev doesn't seem to need an additional variable.